### PR TITLE
feat(notifications): Add additional push APIs

### PR DESCRIPTION
**Pull request by cshfang**
#### Description of changes

This PR adds the rest of the Push Notification APIs refactored as functional. Validation for initialization still to come in a follow up PR.

#### Description of how you validated changes
* Validated the various APIs against Push Notification test app on React Native

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/notifications/src/pushNotifications/types/outputs.ts [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { EventListenerRemover } from '../../eventListeners';
+import {
+	PushNotificationMessage,
+	PushNotificationPermissionStatus,
+} from '../types';
+
+export type PushNotificationGetBadgeCountOutput = number;
```
Comment (HuiSF) [inline packages/notifications/src/pushNotifications/types/outputs.ts:10]: Do we need to alias primitive types, is this for the possibility of making the output a union of multiple types?

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/setBadgeCount.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { PlatformNotSupportedError } from '@aws-amplify/core/internals/utils';
+import { SetBadgeCount } from '../types';
+
+/**
+ * Sets the current badge count (the number next to your app's icon). Setting the badge count to 0 (zero) will remove
```
Comment (HuiSF) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/setBadgeCount.ts:8]: nit:
```suggestion
 * Sets the current badge count (the number on the top right corner of your app's icon). Setting the badge count to 0 (zero) will remove
```

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/requestPermissions.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+ * choose not to request specific permissions. The resulting promise will resolve to true if requested permissions are
+ * granted (or have previously been granted) or false otherwise. Not all specific permissions are supported by platforms
+ * your React Native app can run on but will be safely ignored even on those plaforms. Currently supported permissions:
+ *
+ *   Alert: When set to true, requests the ability to display notifications to the user.
+ *
+ *   Sound: When set to true, requests the ability to play a sound in response to notifications.
+ *
+ *   Badge: When set to true, requests the ability to update the app's badge.
+ *
```
Comment (HuiSF) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/requestPermissions.ts:18]: Should this inline doc have description of the parameters? (same to other APIs inline doc)

```suggestion
 *   * `alert`: When set to true, requests the ability to display notifications to the user.
 *
 *   * `sound`: When set to true, requests the ability to play a sound in response to notifications.
 *
 *   * `badge`: When set to true, requests the ability to update the app's badge.
 *
```

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/requestPermissions.ts [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+import { PlatformNotSupportedError } from '@aws-amplify/core/internals/utils';
+import { RequestPermissions } from '../types';
+
+/**
+ * Requests notification permissions from your user. By default, Amplify requests all supported permissions but you can
+ * choose not to request specific permissions. The resulting promise will resolve to true if requested permissions are
+ * granted (or have previously been granted) or false otherwise. Not all specific permissions are supported by platforms
+ * your React Native app can run on but will be safely ignored even on those plaforms. Currently supported permissions:
```
Comment (HuiSF) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/requestPermissions.ts:11]:
```suggestion
 * your React Native app can run on but will be safely ignored even on those platforms. Currently supported permissions:
```

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/onTokenReceived.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+import {
+	OnTokenReceived,
+	OnTokenReceivedInput,
+	OnTokenReceivedOutput,
+} from '../types';
+
+/**
+ * Registers a listener that will be triggered when a token is received. A token will be received:
+ *   1. On every app launch, including the first install
+ *   2. When a token changes (this may happen if the service invalidates the token for any reason)
```
Comment (HuiSF) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/onTokenReceived.ts:14]: nit: should have another condition - request premission succeeded?

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/getPermissionStatus.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ *   'granted' - Permissions have been granted by the user. No further actions are needed and their app is ready to
+ *   display notifications.
+ *
+ *   'denied' - Permissions have been denied by the user. Further attempts to request permissions will no longer
+ *   trigger a permission dialog. Your app should now either degrade gracefully or prompt your user to grant the
+ *   permissions needed in their device settings.
+ *
+ * @throws platform: {@link PlatformNotSupportedError} - Thrown if called against an unsupported platform. Currently,
+ * only React Native is supported by this API.
+ * @return {Promise<GetPermissionStatusOutput>} a promise resolving to a string representing the current status of user
```
Comment (HuiSF) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/getPermissionStatus.ts:26]: I'm seeing different pattern here - should we use jsdoc convent to note the return type within `{}`, or should use ts embedding the type in the function signature itself...

Not a blocking comment.

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/getPermissionStatus.ts [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+ *   'shouldExplainThenRequest' - It is recommended at this time to provide some context or rationale to the user
+ *   explaining why you want to send them push notifications before requesting for permissions.
+ *
+ *   'granted' - Permissions have been granted by the user. No further actions are needed and their app is ready to
+ *   display notifications.
+ *
+ *   'denied' - Permissions have been denied by the user. Further attempts to request permissions will no longer
+ *   trigger a permission dialog. Your app should now either degrade gracefully or prompt your user to grant the
+ *   permissions needed in their device settings.
+ *
```
Comment (HuiSF) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/getPermissionStatus.ts:23]:
```suggestion
 *   * `'shouldRequest'` - No permissions have been requested yet. It is idiomatic at this time to simply request for
 *   * permissions from the user.
 *
 *   * `'shouldExplainThenRequest'` - It is recommended at this time to provide some context or rationale to the user
 *   explaining why you want to send them push notifications before requesting for permissions.
 *
 *   * `'granted'` - Permissions have been granted by the user. No further actions are needed and their app is ready to
 *   display notifications.
 *
 *   * `'denied'` - Permissions have been denied by the user. Further attempts to request permissions will no longer
 *   trigger a permission dialog. Your app should now either degrade gracefully or prompt your user to grant the
 *   permissions needed in their device settings.
 *
```

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/getBadgeCount.ts [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+/**
+ * Returns the current badge count (the number next to your app's icon). This function is safe to call (but will be
+ * ignored) even when your React Native app is running on platforms where badges are not supported.
+ *
+ * @throws platform: {@link PlatformNotSupportedError} - Thrown if called against an unsupported platform. Currently,
+ * only React Native is supported by this API.
+ * @returns A promise that resolves to a number representing the current count displayed on the app badge.
+ * @example
+ * ```ts
+ * const badgeCount = await getBadgeCount();
```
Comment (jimblanc) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/getBadgeCount.ts:16]:
```suggestion
 * const badgeCount = await getBadgeCount();
 * ```
```

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/getLaunchNotification.ts [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+ *   3. Your app was brought back to the foreground by some other means (e.g. user tapped the app icon)
+ *
+ * @throws platform: {@link PlatformNotSupportedError} - Thrown if called against an unsupported platform. Currently,
+ * only React Native is supported by this API.
+ * @returns {Promise<GetLaunchNotificationOutput>} - a promise resolving to {@link PushNotificationMessage} if there is
+ * a launch notification and `null` otherwise.
+ * @example
+ * ```ts
+ * const launchNotification = await getLaunchNotification();
+ */
```
Comment (jimblanc) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/getLaunchNotification.ts:22]:
```suggestion
 * const launchNotification = await getLaunchNotification();
 * ```
 */
```

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/requestPermissions.ts [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+ * const arePermissionsGranted = await requestPermissions();
+ *
+ * @example
+ * ```ts
+ * // Prevent requesting specific permissions
+ * const arePermissionsGranted = await requestPermissions({
+ *   sound: false,
+ *   badge: false
+ * });
+ */
```
Comment (jimblanc) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/requestPermissions.ts:35]:
```suggestion
 * });
 * ```
 */
```

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/setBadgeCount.ts [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+ * the badge from your app's icon.This function is safe to call (but will be ignored) even when your React Native app
+ * is running on platforms where badges are not supported.
+ *
+ * @throws platform: {@link PlatformNotSupportedError} - Thrown if called against an unsupported platform. Currently,
+ * only React Native is supported by this API.
+ * @example
+ * ```ts
+ * setBadgeCount(42);
```
Comment (jimblanc) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/setBadgeCount.ts:16]:
```suggestion
 * setBadgeCount(42);
 * ```
```

File: packages/notifications/src/pushNotifications/providers/pinpoint/apis/setBadgeCount.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import { PlatformNotSupportedError } from '@aws-amplify/core/internals/utils';
+import { SetBadgeCount } from '../types';
+
+/**
+ * Sets the current badge count (the number next to your app's icon). Setting the badge count to 0 (zero) will remove
+ * the badge from your app's icon.This function is safe to call (but will be ignored) even when your React Native app
```
Comment (jimblanc) [inline packages/notifications/src/pushNotifications/providers/pinpoint/apis/setBadgeCount.ts:9]:
```suggestion
 * the badge from your app's icon. This function is safe to call (but will be ignored) even when your React Native app
```

**Review by jimblanc — COMMENTED** [review]
Looks good! Some nits

**Review by jimblanc — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]

**Review by HuiSF — DISMISSED** [review]

**Review by HuiSF — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]

**Pull request merged**